internal/backends/linuxkms/display/egldisplay.rs

```diff
@@ -0,0 +1,250 @@
+    fn present(&self) -> Result<(), Box<dyn std::error::Error>> {
+        let mut front_buffer = unsafe {
+            self.gbm_surface
+                .lock_front_buffer()
+                .map_err(|e| format!("Error locking gmb surface front buffer: {e}"))?
+        };
+        // TODO: respect modifiers & planes if front_buffer has, use add_planar_framebuffer and fall back to add_framebuffer
+        let fb = self
+            .gbm_device
+            .add_framebuffer(&front_buffer, 24, 32)
+            .map_err(|e| format!("Error adding gbm buffer as framebuffer: {e}"))?;
+        front_buffer
+            .set_userdata(OwnedFramebufferHandle { handle: fb, device: self.drm_device.clone() })
+            .map_err(|e| format!("Error setting userdata on gbm surface front buffer: {e}"))?;
+
+        if let Some(last_buffer) = self.last_buffer.replace(Some(front_buffer)) {
+            self.gbm_device
+                .page_flip(self.crtc, fb, drm::control::PageFlipFlags::EVENT, None)
+                .map_err(|e| format!("Error presenting fb: {e}"))?;
+
+            for event in self.gbm_device.receive_events().unwrap() {
+                if matches!(event, drm::control::Event::PageFlip(..)) {
+                    break;
+                }
+            }
+
+            drop(last_buffer);
+        } else {
+            self.gbm_device
+                .set_crtc(self.crtc, Some(fb), (0, 0), &[self.connector.handle()], Some(self.mode))
+                .map_err(|e| format!("Error presenting fb: {e}"))?;
+        }
+
+        Ok(())
+    }
+}
+
+impl raw_window_handle::HasWindowHandle for EglDisplay {
+    fn window_handle(
+        &self,
+    ) -> Result<raw_window_handle::WindowHandle<'_>, raw_window_handle::HandleError> {
+        let mut gbm_surface_handle = raw_window_handle::GbmWindowHandle::empty();
+        gbm_surface_handle.gbm_surface = self.gbm_surface.as_raw() as _;
+
+        // Safety: This is safe because the handle remains valid; the next rwh release provides `new()` without unsafe.
+        let active_handle = unsafe { raw_window_handle::ActiveHandle::new_unchecked() };
+
+        Ok(unsafe {
+            raw_window_handle::WindowHandle::borrow_raw(
+                raw_window_handle::RawWindowHandle::Gbm(gbm_surface_handle),
+                active_handle,
+            )
+        })
+    }
+}
+
+impl raw_window_handle::HasDisplayHandle for EglDisplay {
+    fn display_handle(
+        &self,
+    ) -> Result<raw_window_handle::DisplayHandle<'_>, raw_window_handle::HandleError> {
+        let mut gbm_display_handle = raw_window_handle::GbmDisplayHandle::empty();
+        gbm_display_handle.gbm_device = self.gbm_device.as_raw() as _;
+
+        Ok(unsafe {
+            raw_window_handle::DisplayHandle::borrow_raw(raw_window_handle::RawDisplayHandle::Gbm(
+                gbm_display_handle,
+            ))
+        })
+    }
+}
+
+pub fn create_egl_display(device_opener: &DeviceOpener) -> Result<EglDisplay, PlatformError> {
```

Comment:
I think the error reporting needs more work.
And a lot of this function can be refactored to just open a display that can be used for software rendering.
So it should be refactored. 

But I have change locally that already does that and will try to polish later.

Reply:
Yes, this function in particular needs to be refactored for software rendering support. I'm in favor of having that.